Dynamic scheduling

nateyazdani opened the pull request:
Merge the feature branch.

Commit:
Create missing bucket for one-day frequency

Commit:
Change positional arguments given by keyword

The functions ``get_model()`` and ``get_plugin()`` from ``statscache.utils``
were being called with their final positional arguments supplied as keyword
arguments. Although it works just fine, it is redundant (``config=config``) and
a bit confusing.

Commit:
Separate plugin loading and initialization

Separate plugin loading and initialization so that the list of available plugins
is fixed at start-up and easily accessible without necessarily causing them to
be initialized. This also allows the list of plugin classes to be accessible
before the Moksha Hub configuration is even available. In addition, plugin
entry-points are now required to either be a specific plugin class or collection
of them. There is an associated commit for statscache_plugins.

Commit:
Rename ``plugin.idx`` to ``plugin.ident``

The ``idx`` property of the ``statscache.plugins.BasePlugin`` is a bit of a
misnomer; it's not really a numerical index, as its name might suggest, but
really a unique string identifier. I've tried to rename it appropriately.

Commit:
Relocate cache buckets to respective producers

Move the caches from the StatsConsumer to the owning StatsProducer. This gives us
greater flexibility in creating new producers, because the central consumer does
not need to be specialized for each new producer.

Commit:
Provide timestamp to handle() method of plugins

The abstract method signature of the method handle() on
statscache.plugins.BasePlugin includes a timestamp, but this is not actually
given when the method is invoked. This change adds that, and a corresponding
change in statscache_plugins corrects the method signature in the plugin
implementations.

Commit:
Simplify time lapse/frequency classes

Removes statscache.utils.{Schedule,Frequency} in favor of
statcache.plugins.Frequency, an extension of datetime.timedelta. This class
contain only the minimal amount of code necessary for our uses.

Commit:
Dynamically generate producers based on plugins

Replaces the static list of available producers with a class factory generator
that determines the frequencies desired by the available plugin classes and then
creates a producer class for each frequency. As part of this, producers are now
responsible for initializing their respective plugins and registering them in the
plugin table for access by the web app. This will facilitate merging the
initialize() method of plugins with the standard __init__(), since a database
session will now be available at the time of plugin instantiation.

Commit:
Update statscache.plugins.Frequency

Changes statscache.plugins.Frequency so that all frequencies are synchronized
at UTC midnight and introduces some associated methods.

Commit:
Synchronize producer/plugin frequencies

The statscache.plugins.Frequency class is repurposed as
statscache.frequency.Frequency to serve as a sort of synchronized timedelta.
The producer factory now initializes all instances of the class so that they
are all synchronized on the same moment, UTC midnight of the current day.
To faciltate this, plugin classes now include a timedelta instance as an
'interval' class attribute and receive the associated Frequency instance in the
__init__() method, which attaches it as a 'frequency' instance attribute.

Commit:
Update scheduling/frequency tests

Update the unit tests for the old statscache.schedule.{Frequency,Schedule} to be
used with the new statscache.frequency.Frequency.

Commit:
Correct incorrect keyword parameter default

Moves time-dependent calls into the method body so that the values change with
time, as they are intended to.

Commit:
Simplify Frequency test case method calls

Replace redundant calls to Frequency methods with a single cast, where
appropriate. This could not be done before because freezegun could not affect
the default keyword-argument values as required for these tests to work.

nateyazdani committed:
Merge pull request #18 from yazman/scheduling

Dynamic scheduling based on plugins

Commit:
Include a default value for frequency epoch

Default the synchronization datetime, or epoch, of statscache.frequency.Frequency
instances to UTC midnight of the current day that the module was instantiated,
making it easier for future uses of the class to get consistent behavior. The
only place where objects of the class are currently created, in
statscache.producer.factory(), is switched to the default value, which results
in no behavioral change.

Commit:
Separate frontend plugin initialization (bugfix)

The frontend and backend of statscache share very little in the nature of how
they initialize and access plugin instances, especially with the recent addition
of the producer factory. In fact, their requirements for initialization are
largely incompatible, and running the frontend in its own process did not work,
as plugin instantiation was tied to producer initialization. This is a bug. To
remedy this, the common plugin table is done away with in favor of totally
separate plugin initialization and management paths for the backend and frontend.
For the most part, all this entailed was implementing plugin initialization and
management directly in the web interface code paths.

rtnpro committed:
Merge pull request #19 from yazman/scheduling_bugfix

Plugin initialization bugfix

rtnpro commented:
@yazman 

I tried `fedmsg-hub` on this branch. I am getting an error in `moksha` during running `fedmsg-hub`:

```
[2015-07-10 20:16:03][moksha.hub   ERROR] Failed to init <generator object factory at 0x7fad22e02c80> producer.
Traceback (most recent call last):
  File "build/bdist.linux-x86_64/egg/moksha/hub/hub.py", line 404, in __init_producers
    producer_obj = producer_class(self)
TypeError: 'generator' object is not callable
```

Could you have a look at it?

rtnpro commented on statscache/producer.py:
```diff
@@ -13,81 +14,94 @@
+        yield StatsProducerAnon
+    return
+
+producers = factory()
```
Is it compatible with `moksha`?

```
Traceback (most recent call last):
  File "build/bdist.linux-x86_64/egg/moksha/hub/hub.py", line 404, in __init_producers
    producer_obj = producer_class(self)
TypeError: 'generator' object is not callable
```

nateyazdani replied:
Yes, it is. But, and this is something I should've stated more clearly, it requires [this change](https://github.com/mokshaproject/moksha/pull/29), which is currently only available direct from the git repo. When @ralphbean returns from his vacation, I can work with him to get the version on PyPI updated (he's the package owner).

rtnpro commented on statscache/producer.py:
```diff
@@ -13,81 +14,94 @@
             session = self.make_session()
             try:
-                plugin.handle(session, bucket)
+                plugin.handle(session, timestamp, cache)
```
Also, can we get rid of the `timestamp` in the plugin's `handle` call and signatures. The vision was for plugins to be stateless, and use the timestamp from fedmsg message.

nateyazdani replied:
Sure, I'll push that change to this branch. You may have seen my [prior response](https://github.com/fedora-infra/statscache_plugins/commit/db2ab8e78cfb55a59c4843b7206406bddff265b1#commitcomment-12106627) about this , where I explained my rationale. Essentially, I thought that plugins were to assume that messages received in `handle()` arrived during the interval ending on `self.frequency.last()`, but now that I understand the statelessness of the interface, I do agree with your point :+1:

Commit:
Remove timestamp from plugin handle() method

Make plugins stateless again by removing the timestamp from
statscache.plugins.handle(). The argument had been reintroduced due to a
misunderstanding.

Commit:
Remove accidental line of commented-out code

A line of code had been previously commented out instead of removed and
forgotten about. This simply removes it.

Commit:
Add Moksha Hub version requirement

New features of statscache require a moksha installation with corresponding
support, or else it won't run at all.

ralphbean committed:
Typofix.

ralphbean committed:
Re-arrange how db tables get created.

ralphbean commented:
:+1: to merge, here.
